Support for partition_cols in to_parquet

anjsudh opened the pull request:
- [x] closes [#23283](https://github.com/pandas-dev/pandas/issues/23283)
- [x] tests passed
- [x] passes git diff upstream/master -u -- "*.py" | flake8 --diff
- [x] whatsnew entry

pep8speaks commented:
Hello @anjsudh! Thanks for submitting the PR.

 - In the file [`pandas/io/parquet.py`](https://github.com/anjsudh/pandas/blob/7c7a7a9fe406fb6cdd15c318cad942b2597f52d9/pandas/io/parquet.py), following                are the PEP8 issues :
> [Line 126:80](https://github.com/anjsudh/pandas/blob/7c7a7a9fe406fb6cdd15c318cad942b2597f52d9/pandas/io/parquet.py#L126): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

WillAyd reviewed (changes requested):
Thanks for the PR! Please make sure that you always have tests first and foremost with any submission though. 

Once added feel free to ping back and can take another look

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -124,9 +124,15 @@ def write(self, df, path, compression='snappy',
-            self.api.parquet.write_table(
-                table, path, compression=compression,
-                coerce_timestamps=coerce_timestamps, **kwargs)
+            if 'partition_cols' in kwargs and \
```
style nitpick: we use parenthesis to wrap long lines instead of a backslash.

TomAugspurger commented on pandas/tests/io/test_parquet.py:
```diff
@@ -474,6 +474,19 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
+    def test_partition_cols_supported(self, pa, df_full):
+        partition_cols =['bool','int']
+        df = df_full
+        path="a.txt"
```
We use temporary files so that tests don't litter the current directory. Something like

```
    with tm.ensure_clean() as path:
        df.to_parquet(path, engine=pa, ...)
```

TomAugspurger commented on pandas/tests/io/test_parquet.py:
```diff
@@ -474,6 +474,19 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
+        import pyarrow.parquet as pq
+        from pyarrow.filesystem import LocalFileSystem
+        fs = LocalFileSystem.get_instance()
+        dataset = pq.ParquetDataset(path, filesystem=fs, metadata_nthreads=16)
```
is this `metadata_nthreads` necessary?

TomAugspurger reviewed (commented):
Could you add a release note in 0.24.0 under enhancements? We'll also need to update the docstring for `fname`. IIUC, this will create a directory, rather than a single file, when `partition_cols` is included.

TomAugspurger commented:
The error in https://travis-ci.org/pandas-dev/pandas/jobs/446310514#L2577 suggests that we'll need to put a minimum pyarrow version to support this behavior. Can you check what version that is, and raise a with an ImportError with a nice error message if the pyarrow is too old?

anjsudh committed:
closes #23283

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -252,7 +258,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     df : DataFrame
     path : string
-        File path
+        File path ( Will be used as Directory path if
```
Can remove the parenthesis, replace with a new sentence.

Rather than "Directory path", say "`root_path`", which is the name used by `pyarrow.parquet.write_to_dataset`.

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -252,7 +258,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     path : string
-        File path
+        File path ( Will be used as Directory path if
+        'partition_cols' is provided as parameter for 'pyarrow' engine).
```
Typically parameter names like 'partition_cols' are placed in single backticks.

TomAugspurger commented on pandas/tests/io/test_parquet.py:
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
+    def test_partition_cols_supported(self, pa_ge_070, df_full):
+        partition_cols = ['bool', 'int']
+        df = df_full
+        path = tempfile.mkdtemp()
```
This can use `with tm.ensure_clean() as path:`. Then you don't have to worry about the `shutil.rmtree` and the imports.

anjsudh replied:
tm.ensure_clean() does not create Temporary Directory. It only creates temporary files. Will change the same in another request

gfyoung replied:
Silly question, but why are we passing in a directory for `df.to_parquet` ?

anjsudh replied:
We are passing the path for the temp directory. pyarrow.write_to_dataset requires a directory to store the partitioned files.

anjsudh replied:
@gfyoung For more details, do refer here : https://arrow.apache.org/docs/python/parquet.html#writing-to-partitioned-datasets

jreback replied:
tempfile.mkdtemp is ok, but pls use this in a with statement so it is cleaned.

TomAugspurger commented on pandas/tests/io/test_parquet.py:
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
+        partition_cols = ['bool', 'int']
+        df = df_full
+        path = tempfile.mkdtemp()
+        # supported in >= 0.7.0
```
Can remove this comment, since we have the `pa_ge_070` fixture.

TomAugspurger commented on pandas/tests/io/test_parquet.py:
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
+                      compression=None)
+        import pyarrow.parquet as pq
+        dataset = pq.ParquetDataset(path, validate_schema=False)
+        assert len(dataset.pieces) > 0
```
Can we make this assert more precise? Do we know that it will be equal to the number of combinations in the `['bool', 'int']` columns?

TomAugspurger commented on pandas/tests/io/test_parquet.py:
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
+    def test_ignore_partition_cols_lt_070(self, pa_lt_070, df_full):
+        partition_cols = ['bool', 'int']
+        df = df_full
+        # supported in >= 0.7.0
```
Can remove this comment.

TomAugspurger reviewed (commented):
We should mention somewhere that

anjsudh committed:
Fix linting issue

anjsudh commented:
@WillAyd Hi, hope you can review the diff now? Have added the necessary tests

xhochy commented:
From a `pyarrow` perspective this is LGTM.

jreback commented on pandas/io/parquet.py:
```diff
@@ -252,7 +257,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     ----------
     df : DataFrame
     path : string
-        File path
```
I am not sure this doc-string is formatted correctly. @datapythonista ?

also pls list the ``partition_cols`` arg with a versionadded tag (and provide a description).

this only applies to pyarrow? or also fastparquet?

anjsudh replied:
this applies only for pyarrow

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
+        assert dataset.partitions.partition_names == set(partition_cols)
+        shutil.rmtree(path)
+
+    def test_ignore_partition_cols_lt_070(self, pa_lt_070, df_full):
```
add the issue number as a comment

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -478,6 +480,25 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
+        df = df_full
+        path = tempfile.mkdtemp()
+        df.to_parquet(path, partition_cols=partition_cols,
+                      compression=None)
```
is there a reason you are also not reading this back with ``pd.read_parquet``?

anjsudh replied:
The objective of this test was to to ensure the file is partitioned. Hence covered that alone in this test. I am not clear on how to get the number of partitions in the file using read_parquet

jreback reviewed (changes requested):
also pls update the documentatio in io.rst

datapythonista commented on pandas/io/parquet.py:
```diff
@@ -252,7 +257,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     ----------
     df : DataFrame
```
Can you remove this parameter please. We don't document the self.

datapythonista commented on pandas/io/parquet.py:
```diff
@@ -252,7 +257,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     ----------
     df : DataFrame
     path : string
```
Can you replace `string` by the Python type `str`?

datapythonista commented on pandas/io/parquet.py:
```diff
@@ -252,7 +257,8 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
     path : string
-        File path
+        File path ( Will be used as `root_path` if
+        `partition_cols` is provided as parameter for 'pyarrow' engine).
```
Can you rephrase this please. It's not very clear what you mean.

datapythonista reviewed (commented):
Suggested couple of improvements to the docstring. Also, if you can run `./scripts/validate_docstrings.py pandas.DataFrame.to_parquet` to see that everything else is correct in it, that would be great.

anjsudh committed:
Update documentation
using with to clean directory
support for fastparquet

anjsudh commented:
closes #23373 in a way

TomAugspurger commented on pandas/util/testing.py:
```diff
@@ -772,6 +772,23 @@ def ensure_clean(filename=None, return_filelike=False):
+        yield directory_name
+    finally:
+        import shutil
+        shutil.rmtree(directory_name)
```
This should be in a `try / except Exception`, since we shouldn't ever fail in a test cleanup.

TomAugspurger commented on pandas/util/testing.py:
```diff
@@ -772,6 +772,23 @@ def ensure_clean(filename=None, return_filelike=False):
+    Get a temporary directory path and agrees to remove on close.
+
+    Yields
+    ----------
```
Underline should match the length of Yields

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -266,11 +278,18 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
+            Column names by which to partition the dataset
+            Columns are partitioned in the order they are given
+            The behaviour applies only to pyarrow >= 0.7.0 and fastparquet
+            For other versions, this argument will be ignored.
```
Is it ignored, or doesn't it pass through to `write_table` and raise?

I would prefer that we raise for old pyarrows, so that a user doesn't think this is working when it isn't.

Also, I thought that fast parquet called it `partition_on`.

anjsudh replied:
Yes, we will be passing this key as partition_on

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -224,7 +235,8 @@ def write(self, df, path, compression='snappy', index=None, **kwargs):
         with catch_warnings(record=True):
             self.api.write(path, df, compression=compression,
-                           write_index=index, **kwargs)
+                           write_index=index, partition_on=partition_cols,
```
What if a user does

```python
df.to_parquet("/tmp/a", engine="fastparquet", partition_on=['foo'])
```

I think that would raise an error, since you've specified `partition_on` twice. It should use the user-provided `partition_on`.

anjsudh replied:
i feel using one argument for partitioning makes more sense. Why don't we raise an error, asking them to use partition_cols instead in that case

anjsudh replied:
make sense.. will handle it such that partition_on gets more priority than partition_cols with a warning

TomAugspurger replied:
I'm not sure it should even be a warning. If a warning is going to be emitted, then that should be done by fastparquet.

jreback commented on pandas/core/frame.py:
```diff
@@ -1998,6 +1999,12 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
+            Columns are partitioned in the order they are given
+            The behaviour applies only to pyarrow >= 0.7.0 and fastparquet
+            For other versions, this argument will be ignored.
+            .. versionadded:: 0.24.0
```
need a blank line before the versionadded. we should prob just bump our min of pyarrow to 0.8.0 anyhow, can you create a new issue for this.

jreback commented on pandas/io/parquet.py:
```diff
@@ -206,12 +213,16 @@ def __init__(self):
         # Use tobytes() instead.
 
+        if partition_cols is not None:
+            kwargs['file_scheme'] = 'hive'
```
is this not defaulted pyarrow? if you don't specify what is the default?

anjsudh replied:
this is for fastparquet. Default file_scheme is simple. In simple _metadata file will not be generated. _metadata file is necessary to perform a read on a directory in fastparquet

martindurant replied:
To be specific, simple means one file only, and is not appropriate for dask. There is also Drill scheme, where the directory names are different from the Hive variant (*if* using directory partitioning, no difference otherwise). Spark defaults to the Hive format, it's a reasonable default in every case.

martindurant replied:
(the _metadata file is longer necessary for the fastparquet branch, actually, but it is preferred)

jreback replied:
can you add a comment above this describing what you said above, e.g. for compat with fp

TomAugspurger replied:
Need a comment here sayings what's going on.

martindurant replied:
That is expected in 0.1.6, you would need to pass the whole list of data files rather than the root directory. This may be fixed in master.

anjsudh committed:
Documentation changes
Throws value error for older versions of pyarrow and if partitions_on is used in fastparquet

TomAugspurger commented:
cc @martindurant for fastparquet things:

tldr: this PR is adding support for writing a dataframe partitioned by columns with the pyarrow engine.

1. fastparquet uses `partition_on`, pyarrow uses `partition_cols`. @xhochy @martindurant Any thoughts on standardizing on a name (stepping back, these do the same thing, correct?)
2. @martindurant thoughts on the change at 
https://github.com/pandas-dev/pandas/pull/23321#discussion_r228909418?

anjsudh committed:
not raise error when partion_on is used in fastparquet
add a blank line before the versionadded

anjsudh commented:
@WillAyd @jreback @TomAugspurger Have made the necessary changes. Hope you can have a look?

jreback commented on pandas/io/parquet.py:
```diff
@@ -206,12 +217,20 @@ def __init__(self):
         # Use tobytes() instead.
 
+        if 'partition_on' in kwargs:
+            partition_cols = kwargs['partition_on']
```
use .pop()

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -543,3 +565,27 @@ def test_s3_roundtrip(self, df_compat, s3_resource, fp):
+            actual_partition_cols = fastparquet.ParquetFile(path, False).cats
+            assert len(actual_partition_cols) == 2
+
+    def test_partition_on_supported(self, fp, df_full):
```
what happens with a non-list-like partition column? does this need checking?

anjsudh replied:
both fastparquet and pyarrow require a list as input. So even if only one column is required, it is given as a list.

jreback commented on pandas/tests/util/test_testing.py:
```diff
@@ -875,3 +875,12 @@ def test_datapath_missing(datapath, request):
+        assert os.path.exists(path)
+        assert os.path.isdir(path)
+        temppath = path
+    assert not os.path.exists(temppath)
```
you can just use path here, its still in scope

jreback commented on pandas/util/testing.py:
```diff
@@ -772,6 +772,26 @@ def ensure_clean(filename=None, return_filelike=False):
+        yield directory_name
+    finally:
+        try:
+            import shutil
```
this import can be at the top of the file

jreback commented on pandas/util/testing.py:
```diff
@@ -772,6 +772,26 @@ def ensure_clean(filename=None, return_filelike=False):
+            import shutil
+            shutil.rmtree(directory_name)
+        except Exception as e:
+            print("Exception on removing folder: {error}".format(error=e))
```
just pass

anjsudh committed:
using pop for array element deletion
passing exception during folder removal
removal of additional variable

anjsudh committed:
using pop for array element deletion
passing exception during folder removal
removal of additional variable

anjsudh committed:
Merge branch 'master' into master

anjsudh committed:
fix order of imports

anjsudh committed:
Merge remote-tracking branch 'upstream/master'

anjsudh committed:
Merge remote-tracking branch 'upstream/master'

anjsudh committed:
fix failing codecheck

TomAugspurger commented on doc/source/io.rst:
```diff
@@ -4574,6 +4574,8 @@ Several caveats.
 * Categorical dtypes can be serialized to parquet, but will de-serialize as ``object`` dtype.
 * Non supported types include ``Period`` and actual Python object types. These will raise a helpful error message
   on an attempt at serialization.
+* ``partition_cols`` will be used for partitioning the dataset, where the dataset will be written to multiple
```
The rest of the items in this lists feel more like limitations of pandas / these engines. Requiring that `path` be a directory when `partition_cols` is set doesn't seem to fit here.

I think this is important / different enough to deserve a new small section below "Handling Indexes", with

1. A description of what `partition_cols` requires (list of column names, directory for file path)
2. A description of why you might want to use partition_cols
3. A small example.

anjsudh replied:
done

TomAugspurger commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -235,6 +235,7 @@ Other Enhancements
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
 - Added :meth:`Interval.overlaps`, :meth:`IntervalArray.overlaps`, and :meth:`IntervalIndex.overlaps` for determining overlaps between interval-like objects (:issue:`21998`)
+- :func:`~DataFrame.to_parquet` now supports writing a DataFrame as a directory of parquet files partitioned by a subset of the columns. (:issue:`23283`).
```
Probably best to mention "with the pyarrow engine (this was previously supported with fastparquet)."

anjsudh replied:
done

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -1997,6 +1998,13 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
+            Column names by which to partition the dataset
+            Columns are partitioned in the order they are given
+            The behaviour applies only to pyarrow >= 0.7.0 and fastparquet
+            For other versions, this argument will be ignored.
```
Is it actually ignored for older pyarrows? I would have hoped it would raise when pyarrow gets the unrecognized argument.

TomAugspurger replied:
Actually it seems like we raise. Could you update this?

anjsudh replied:
done

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -208,12 +219,19 @@ def __init__(self):
         # Use tobytes() instead.
 
+        if 'partition_on' in kwargs:
+            partition_cols = kwargs.pop('partition_on')
```
Should check that `partition_cols` is None before doing this. We would want to raise on

```python
df.to_parquet(engine='fastparquet', partition_cols=['a', 'b'], partition_on=['b', 'c'])
```

anjsudh replied:
done

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -268,11 +287,18 @@ def to_parquet(df, path, engine='auto', compression='snappy', index=None,
+            Column names by which to partition the dataset
+            Columns are partitioned in the order they are given
+            The behaviour applies only to pyarrow >= 0.7.0 and fastparquet
+            For other versions, this argument will be ignored.
```
Update this: We don't ignore.

anjsudh replied:
missed updating the docs.. thanks :)

anjsudh committed:
Updated documentation
Fix"Should raise error on using partition_cols and partition_on together"

TomAugspurger commented:
Merge conflict in `io/parquet.py`, if you could fix that up I think this will be good to go.

anjsudh committed:
Merge branch 'master' into master

anjsudh committed:
Removed < 0.7.0 documentation for pyarrow in partition support code

jreback commented on pandas/io/parquet.py:
```diff
@@ -194,15 +213,15 @@ def read(self, path, columns=None, **kwargs):
-        File path
+    path : str
+        File path or Root Directory path. Will be used as Root Directory path
+        while writing a partitioned dataset.
```
versionchanged tag here

anjsudh replied:
done

anjsudh committed:
documentation changes for version change

anjsudh commented on pandas/core/frame.py:
```diff
@@ -1984,7 +1984,8 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
         fname : str
-            String file path.
+            File path or Root Directory path. Will be used as Root Directory
+            path while writing a partitioned dataset.
```
done

anjsudh commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -235,6 +235,7 @@ Other Enhancements
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
 - Added :meth:`Interval.overlaps`, :meth:`IntervalArray.overlaps`, and :meth:`IntervalIndex.overlaps` for determining overlaps between interval-like objects (:issue:`21998`)
+- With the pyarrow engine, :func:`~DataFrame.to_parquet` now supports writing a DataFrame as a directory of parquet files partitioned by a subset of the columns. (:issue:`23283`).
```
done

anjsudh commented on doc/source/io.rst:
```diff
@@ -4668,6 +4668,33 @@ Passing ``index=True`` will *always* write the index, even if that's not the
+The `partition_cols` are the column names by which the dataset will be partitioned.
+Columns are partitioned in the order they are given. The partition splits are
+determined by the unique values in the partition columns.
+The above example creates a partitioned dataset that may look like:
```
done

anjsudh commented on doc/source/io.rst:
```diff
@@ -4668,6 +4668,33 @@ Passing ``index=True`` will *always* write the index, even if that's not the
 
 
+Partitioning Parquet files
+''''''''''''''''''''''''''
```
done

jreback commented on doc/source/io.rst:
```diff
@@ -4668,6 +4668,36 @@ Passing ``index=True`` will *always* write the index, even if that's not the
+Parquet supports partitioning of data based on the values of one or more columns.
+
+.. ipython:: python
+   :suppress:
```
don't add the suppress here (this won't print the block). Rather add another block after than cleans up the dir that is created (and use suppress on that)

anjsudh committed:
Cleanup file in while generating doc

anjsudh commented:
@jreback @WillAyd hope you can have a look ?

jreback commented on pandas/core/frame.py:
```diff
@@ -1984,7 +1984,10 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
         Parameters
         ----------
         fname : str
```
side issue. we use ``path`` elsewhere for IO routines. We should change this as well (out of scope here). would have to deprecate (the name) unfortunately.

jreback replied:
we actually use path on the top-level ``.to_parquet``, not sure how this is named this way.

jreback reviewed (approved):
off topic comments, lgtm. @WillAyd

WillAyd commented on doc/source/io.rst:
```diff
@@ -4668,6 +4668,43 @@ Passing ``index=True`` will *always* write the index, even if that's not the
+   from shutil import rmtree
+   try:
+       rmtree('test')
+   except Exception:
```
Can we be more specific about the exception here? I believe this could just be an `OSError` or `FileNotFoundError` depending on version. Would rather catch those so that this doesn't end up catching things it doesn't mean to during the doc build process

WillAyd replied:
Hmm OK. No concern around it catching exceptions that it shouldn't though?

WillAyd commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -235,6 +235,7 @@ Other Enhancements
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
 - Added :meth:`Interval.overlaps`, :meth:`IntervalArray.overlaps`, and :meth:`IntervalIndex.overlaps` for determining overlaps between interval-like objects (:issue:`21998`)
+- :func:`~DataFrame.to_parquet` now supports writing a ``DataFrame`` as a directory of parquet files partitioned by a subset of the columns, when ``engine = 'pyarrow'``. (:issue:`23283`).
```
Very minor nits but you don't need the comma after "subset of the columns" nor do you need the two periods towards the end

anjsudh replied:
Done

WillAyd commented on pandas/core/frame.py:
```diff
@@ -1984,7 +1984,10 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
+            File path or Root Directory path. Will be used as Root Directory
+            path while writing a partitioned dataset.
+
+            .. versionchanged:: 0.24.0
```
Does this render? Might need a blank space after the directive?

anjsudh replied:
Looks like it does not.

Documentation shows API Reference for DataFrame.to_parquet, read_parquet() for parquet.py and io.rst, which explains about reading and writing to parquet.

WillAyd replied:
Wasn't referring to the public API as much as the docstring in isolation. If you can ensure there is a blank line after this and any of the other `versionadded` directives should be the last thing

anjsudh replied:
have added an empty line after all versionchanged and versionadded. Hope you can have a look

WillAyd commented on pandas/core/frame.py:
```diff
@@ -1997,6 +2000,11 @@ def to_parquet(self, fname, engine='auto', compression='snappy',
             If ``False``, they will not be written to the file. If ``None``,
             the behavior depends on the chosen engine.
 
+            .. versionadded:: 0.24.0
```
Same question as above

WillAyd commented on pandas/io/parquet.py:
```diff
@@ -156,12 +162,24 @@ def __init__(self):
         # Use tobytes() instead.
 
+        if 'partition_on' in kwargs:
+            if partition_cols is None:
```
Let's simplify this as `if 'partition_on' in kwargs and partition_cols is None`

anjsudh replied:
Done

TomAugspurger commented:
Nope. I don't really care about bugs in shutil :)

On Thu, Nov 8, 2018 at 10:31 AM William Ayd <notifications@github.com>
wrote:

> *@WillAyd* commented on this pull request.
> ------------------------------
>
> In doc/source/io.rst
> <https://github.com/pandas-dev/pandas/pull/23321#discussion_r231966117>:
>
> > +
> +    test
> +    ├── a=0
> +    │   ├── 0bac803e32dc42ae83fddfd029cbdebc.parquet
> +    │   └──  ...
> +    └── a=1
> +        ├── e6ab24a4f45147b49b54a662f0c412a3.parquet
> +        └── ...
> +
> +.. ipython:: python
> +   :suppress:
> +
> +   from shutil import rmtree
> +   try:
> +       rmtree('test')
> +   except Exception:
>
> Hmm OK. No concern around it catching exceptions that it shouldn't though?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/23321#discussion_r231966117>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIjUauolVWoYkrRq48bKY5stxCe4eks5utFxdgaJpZM4X4g_n>
> .
>

anjsudh committed:
Text changes, Style changes

anjsudh committed:
added empty line after versionadded

TomAugspurger commented:
Looks good. Nice work @anjsudh!
